chips/sam4l/src/wdt.rs

```diff
@@ -93,86 +95,127 @@ register_bitfields![u32,
 ];
 
 // Page 59 of SAM4L data sheet
-const BASE_ADDRESS: *mut WdtRegisters = 0x400F0C00 as *mut WdtRegisters;
+const WDT_BASE: *mut WdtRegisters = 0x400F0C00 as *mut WdtRegisters;
+const WDT_REGS: StaticRef<WdtRegisters> =
+    unsafe { StaticRef::new(WDT_BASE as *const WdtRegisters) };
 
 pub struct Wdt {
-    registers: *mut WdtRegisters,
     enabled: Cell<bool>,
 }
 
-pub static mut WDT: Wdt = Wdt::new(BASE_ADDRESS);
+pub static mut WDT: Wdt = Wdt::new();
+
+#[derive(Copy, Clone)]
+pub enum WdtClockSource {
+    ClockRCSys = 0,
+    ClockOsc32 = 1,
+}
+
+impl From<WdtClockSource> for FieldValue<u32, Control::Register> {
+    fn from(clock: WdtClockSource) -> Self {
+        match clock {
+            WdtClockSource::ClockRCSys => Control::CSSEL::RCSYS,
+            WdtClockSource::ClockOsc32 => Control::CSSEL::OSC32K,
+        }
+    }
+}
 
 impl Wdt {
-    const fn new(base_address: *mut WdtRegisters) -> Wdt {
+    const fn new() -> Wdt {
         Wdt {
-            registers: base_address,
             enabled: Cell::new(false),
         }
     }
 
-    fn start(&self, period: usize) {
-        let regs: &WdtRegisters = unsafe { &*self.registers };
+    /// WDT Errata: §45.1.3
+    ///
+    /// When writing any of the PSEL, TBAN, EN, or MODE fields, must insert a
+    /// delay for synchronization to complete.
+    ///
+    /// Also handle the KEY for the caller since we're special casing this.
+    fn write_cr(&self, control: FieldValue<u32, Control::Register>) {
+        WDT_REGS.cr.modify(Control::KEY::KEY1 + control);
+        WDT_REGS.cr.modify(Control::KEY::KEY2 + control);
+
+        // When writing to the affected fields, the user must ensure a wait
+        // corresponding to 2 clock cycles of both the WDT peripheral bus clock
+        // and the selected WDT clock source.
+        //
+        // TODO: Actual math based on chosen clock, ASF does:
+        //       delay = div_ceil(sysclk_hz(), OSC_[chosen]_NOMINAL_HZ)
+        unsafe {
+            for _ in 0..10000 {
+                asm!("nop");
```

Comment:
we have a `kernel::support::nop()` function which is marked safe, so you can use that and get rid of the `unsafe`.